src/backend/cpu/TNJ/BinaryNode.hpp

```diff
@@ -44,14 +44,24 @@ namespace TNJ
         {
         }
 
-        void calc(int x, int y, int z, int w)
+        void calc(int x, int y, int z, int w, int lim)
         {
-            this->m_val = m_op.eval(m_lhs->m_val, m_rhs->m_val);
+            Ti *lhs_ptr = &(m_lhs->m_val.front());
+            Ti *rhs_ptr = &(m_rhs->m_val.front());
+            To *out_ptr = &(this->m_val.front());
+            for (int i = 0; i < lim; i++) {
+                out_ptr[i] = m_op.eval(lhs_ptr[i], rhs_ptr[i]);
+            }
         }
 
-        void calc(int idx)
+        void calc(int idx, int lim)
         {
-            this->m_val = m_op.eval(m_lhs->m_val, m_rhs->m_val);
+            Ti *lhs_ptr = &(m_lhs->m_val.front());
+            Ti *rhs_ptr = &(m_rhs->m_val.front());
+            To *out_ptr = &(this->m_val.front());
+            for (int i = 0; i < lim; i++) {
+                out_ptr[i] = m_op.eval(lhs_ptr[i], rhs_ptr[i]);
+            }
```

Comment:
We may convert these structures to Functors in CUDA/OpenCL backends if applicable but it may not help as much as it does in CPU backend I believe.